File: corpus/19_parser.py
import sys

def lex(s):
    s = s.replace("(", " ( ").replace(")", " ) ")
    return [float(w) if w[0].isdigit() else w for w in s.split()]

def expr(ts):
    v, r = term(ts)
    while r and r[0] in "+-":
        v2, r2 = term(r[1:])
        v = v + v2 if r[0] == "+" else v - v2
        r = r2
    return v, r

def term(ts):
    v, r = factor(ts)
    while r and r[0] in "*/":
        v2, r2 = factor(r[1:])
        v = v * v2 if r[0] == "*" else v / v2
        r = r2
    return v, r

def factor(ts):
    if isinstance(ts[0], float):
        return ts[0], ts[1:]
    v, r = expr(ts[1:])
    return v, r[1:]

print(expr(lex(sys.argv[1]))[0])

File: corpus/19_parser.rs
use std::env;

#[derive(Clone, PartialEq)]
enum Tok {
    Num(f64),
    Sym(String),
}

fn lex(s: &str) -> Vec<Tok> {
    let s = s.replace('(', " ( ").replace(')', " ) ");
    s.split_whitespace()
        .map(|w| match w.parse::<f64>() {
            Ok(n) => Tok::Num(n),
            Err(_) => Tok::Sym(w.to_string()),
        })
        .collect()
}

fn expr(ts: &[Tok]) -> (f64, &[Tok]) {
    let (mut v, mut r) = term(ts);
    while let Some(Tok::Sym(op)) = r.first() {
        if op != "+" && op != "-" {
            break;
        }
        let (v2, r2) = term(&r[1..]);
        v = if op == "+" { v + v2 } else { v - v2 };
        r = r2;
    }
    (v, r)
}

fn term(ts: &[Tok]) -> (f64, &[Tok]) {
    let (mut v, mut r) = factor(ts);
    while let Some(Tok::Sym(op)) = r.first() {
        if op != "*" && op != "/" {
            break;
        }
        let (v2, r2) = factor(&r[1..]);
        v = if op == "*" { v * v2 } else { v / v2 };
        r = r2;
    }
    (v, r)
}

fn factor(ts: &[Tok]) -> (f64, &[Tok]) {
    match &ts[0] {
        Tok::Num(n) => (*n, &ts[1..]),
        Tok::Sym(_) => {
            let (v, r) = expr(&ts[1..]);
            (v, &r[1..])
        }
    }
}

fn main() {
    let arg = env::args().nth(1).unwrap();
    let (v, _) = expr(&lex(&arg));
    println!("{v}");
}
